### PR TITLE
SOLR-13101: Concurrency tests for SHARED collection.

### PR DESCRIPTION
**Pull request by mbwaheed**
-Concurrency tests for SHARED collection.
-On core container shutdown, shutdown the background blob delete manager too.
-Use correct casing for SharedCoreStage enum.
-Added INDEXING_BATCH_FINISHED to indicate the end of a batch (BLOB_PUSH_FINISHED does not help if push itself runs into error)

### CODE REVIEW

File: solr/core/src/test/org/apache/solr/store/shared/SharedCoreConcurrencyTest.java [hunk 1]
```diff
@@ -16,31 +16,40 @@
  */
```
Comment (mbwaheed) [inline solr/core/src/test/org/apache/solr/store/shared/SharedCoreConcurrencyTest.java]: Will be easier to review this as a new class, as most of old logic is changed.

File: solr/core/src/test/org/apache/solr/store/shared/SharedCoreConcurrencyTest.java [hunk 2]
```diff
@@ -49,411 +58,673 @@
...
   }
 
-  private void configureTestSharedConcurrencyControllerForNode(JettySolrRunner runner, List<String> progress) {
-    SharedCoreConcurrencyController concurrencyController = new SharedCoreConcurrencyController(runner.getCoreContainer()) {
-      Object recorderLock = new Object();
-      @Override
-      public void recordState(String collectionName, String shardName, String coreName, SharedCoreStage stage) {
-        super.recordState(collectionName, shardName, coreName, stage);
+  /**
+   * Directly query the {@code replica} and verify that all the docs indexed(after accounting for deletions) are found.
```
Comment (andyvuong) [inline solr/core/src/test/org/apache/solr/store/shared/SharedCoreConcurrencyTest.java]: Might be worth adding a note that this is testing an empty replica hydrating after the concurrent tests

File: solr/core/src/test/org/apache/solr/store/shared/SharedCoreConcurrencyTest.java [hunk 3]
```diff
@@ -49,411 +58,673 @@
...
+      for (int i = 1; i <= testState.numDocsIndexed.get(); i++) {
+        String docId = Integer.toString(i);
+        if (!testState.idsDeleted.contains(docId)) {
+          docsExpected.add(docId);
+        }
       }
-    };
-    setupTestSharedConcurrencyControllerForNode(concurrencyController, runner);
+      log.info("numDocsFound=" + docs.size() + " docsFound= " + docs.toString());
+      assertEquals("wrong docs", docsExpected.size() + docsExpected.toString(), docs.size() + docs.toString());
```
Comment (andyvuong) [inline solr/core/src/test/org/apache/solr/store/shared/SharedCoreConcurrencyTest.java]: docsExpected.size() + docsExpected.toString() intentional? Why not two separate assert statements

File: solr/core/src/test/org/apache/solr/store/shared/SharedCoreConcurrencyTest.java [hunk 4]
```diff
@@ -49,411 +58,673 @@
...
-      FileUtils.cleanDirectory(sharedStoreRootPath.toFile());
     }
   }
 
-  private void assertCriticalSections(List<String> progress) {
-    String currentThreadId = null;
+  /**
+   * Analyze core's concurrency stages to make sure no critical section was breached.
+   */
+  private void analyzeCoreConcurrencyStagesForBreaches(String coreName, List<String> coreConcurrencyStages) {
```
Comment (andyvuong) [inline solr/core/src/test/org/apache/solr/store/shared/SharedCoreConcurrencyTest.java]: A note pointing to SharedCoreConcurrencyController here might help future devs understand the implied contracts we're testing here

File: solr/core/src/test/org/apache/solr/store/shared/SharedCoreConcurrencyTest.java (hunk 3 above)
Comment (mbwaheed) [reply solr/core/src/test/org/apache/solr/store/shared/SharedCoreConcurrencyTest.java]: Yes. It helps in debugging to see what docs are different. You can go in the earlier log lines and correlate to exact indexing batch.

File: solr/core/src/test/org/apache/solr/store/shared/SharedCoreConcurrencyTest.java (hunk 4 above)
Comment (mbwaheed) [reply solr/core/src/test/org/apache/solr/store/shared/SharedCoreConcurrencyTest.java]: Updated comment.

File: solr/core/src/test/org/apache/solr/store/shared/SharedCoreConcurrencyTest.java [hunk 5]
```diff
@@ -49,411 +58,673 @@
...
+  }
 
-    for (int i = 0; i < numIndexingThreads; i++) {
-      indexingThreads[i].start();
-    }
+  /**
+   * Sends update request to the server, randomly choosing whether to send it with commit=true or not
+   */
+  private void processUpdateRequest(UpdateRequest request) throws Exception {
+    UpdateResponse response = random().nextBoolean()
```
[the remark this answers is not in the document]
Comment (mbwaheed) [reply solr/core/src/test/org/apache/solr/store/shared/SharedCoreConcurrencyTest.java]: Updated comment.
